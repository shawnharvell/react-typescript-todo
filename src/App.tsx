import React, { useState } from "react";
import { initialTodos } from "./initialTodos";
import { TodoList } from "./TodoList";
import { AddTodoForm } from "./AddTodoForm";
import { Todo, ToggleComplete, AddTodo } from "./types";

const App: React.FC = () => {
  const [todos, setTodos] = useState<Array<Todo>>(initialTodos);

  const toggleComplete: ToggleComplete = selectedTodo => {
    const updatedTodos = todos.map(todo => {
      if (todo === selectedTodo) {
        return { ...todo, complete: !todo.complete };
      }
      return todo;
    });
    setTodos(updatedTodos);
  };

  const addTodo: AddTodo = newTodo => {
    newTodo.trim() !== "" &&
      setTodos([...todos, { text: newTodo, complete: false }]);
  };

  return (
    <div className="app-container">
      <TodoList todos={todos} toggleComplete={toggleComplete} />
      <AddTodoForm addTodo={addTodo} />
    </div>
  );
};

export default App;
