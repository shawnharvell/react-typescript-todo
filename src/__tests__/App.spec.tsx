import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import App from "../App";

describe("App", () => {
  it("snapshot test", () => {
    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });

  it("user interactions", async () => {
    render(<App />);

    expect(screen.queryByPlaceholderText("I need to do...")).toHaveFocus();
    userEvent.type(screen.getByPlaceholderText("I need to do..."), "My New Todo");
    expect(screen.queryByDisplayValue("My New Todo")).toBeInTheDocument();
    userEvent.click(screen.getByText("Add"));
    expect(screen.queryByText("My New Todo")).toBeInTheDocument();

    expect(screen.getByPlaceholderText("I need to do...")).toHaveFocus();
    expect(screen.getByPlaceholderText("I need to do...")).toHaveValue("");

    userEvent.type(screen.getByPlaceholderText("I need to do..."), "My 2nd Todo{enter}");
    expect(screen.queryByText("My 2nd Todo")).toBeInTheDocument();

    expect(screen.getByPlaceholderText("I need to do...")).toHaveFocus();
    expect(screen.getByPlaceholderText("I need to do...")).toHaveValue("");

    userEvent.type(screen.getByPlaceholderText("I need to do..."), "My 3rd Todo{esc}");
    expect(screen.queryByText("My 3rd Todo")).not.toBeInTheDocument();

    expect(screen.getByLabelText("My 2nd Todo").parentElement).not.toHaveClass("complete");
    userEvent.click(screen.getByLabelText("My 2nd Todo"));
    expect(screen.getByLabelText("My 2nd Todo").parentElement).toHaveClass("complete");
    await waitFor(() =>
      expect(screen.getByLabelText("My 2nd Todo").parentElement).toHaveClass("complete")
    );
  });
});
