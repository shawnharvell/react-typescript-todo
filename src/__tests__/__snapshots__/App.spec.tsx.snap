// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot test 1`] = `
<body>
  <div>
    <div
      class="app-container"
    >
      <div
        class="todo-list-container"
      >
        <ul>
          <li>
            <label
              class="complete"
            >
              <input
                checked=""
                type="checkbox"
              />
              Walk the dog
            </label>
          </li>
          <li>
            <label>
              <input
                type="checkbox"
              />
              Make app
            </label>
          </li>
        </ul>
      </div>
      <div
        class="add-form-container"
      >
        <div
          class="textfield"
        >
          <input
            data-testid="new-todo-textbox"
            placeholder="I need to do..."
            type="text"
            value=""
          />
        </div>
        <div
          class="button"
        >
          <button>
            Add
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
